### PR TITLE
Closes #8502: Parallel NDJSON file reading

### LINKED ISSUE #8502: Parallel NDSON file reading
### Is your feature request related to a problem or challenge?

DataFusion can now automatically read CSV and parquet files in parallel (see https://github.com/apache/arrow-datafusion/issues/6325 for CSV)

It would be great to do the same for "NDJSON" files -- namely files that have multiple JSON objects placed one after the other. 

### Describe the solution you'd like

Basically implement what is described in https://github.com/apache/arrow-datafusion/issues/6325 for JSON -- and read a single large ND json file (new line delimited file) in parallel

### Describe alternatives you've considered

Some research may be required -- I am not sure if finding record boundaries is feasible

### Additional context

I found this while writing tests for https://github.com/apache/arrow-datafusion/issues/8451

### PR DESCRIPTION
**Pull request by marvinlanhenke**
## Which issue does this PR close?

Closes #8502.

## Rationale for this change

As stated in the issue: 

> DataFusion can now automatically read CSV and parquet files in parallel (see https://github.com/apache/arrow-datafusion/issues/6325 for CSV)
> It would be great to do the same for "NDJSON" files -- namely files that have multiple JSON objects placed one after the other.

## What changes are included in this PR?

Basically the same approach as in #6325.
However, I refactored and extracted some of the common functions like `find_first_newline` which is used by `calculate_range`, to be used by both the CSV and JSON implementation.

## Are these changes tested?

Yes, basic tests are included.
However, I was not sure about benchmarking the changes (since `benchmarks/bench.sh` does not provide JSON dataset?).

## Are there any user-facing changes?
No.

### CODE REVIEW

**Comment by alamb** [comment]
Thank you for this @marvinlanhenke  - I plan to review it tomorrow.

cc @devinjdangelo

**Comment by marvinlanhenke** [comment]
...just as a sidenote and originally stated in #6801 and #6922; this implementation is suboptimal since it performs multiple get operations on the object store. This will be handled in a follow-up Issue / PR - which I will (most) likely tackle in the next week.

File: datafusion/core/src/datasource/file_format/json.rs [hunk 1]
```diff
@@ -441,4 +446,94 @@ mod tests {
             .collect::<Vec<_>>();
         assert_eq!(vec!["a: Int64", "b: Float64", "c: Boolean"], fields);
     }
+
+    async fn count_num_partitions(ctx: &SessionContext, query: &str) -> Result<usize> {
```
Comment (alamb) [inline datafusion/core/src/datasource/file_format/json.rs:450]: Another potential way to figure out the file groups would be to make the physical plan and then walk it to find the `JsonExec` and its number of partitions

**Comment by marvinlanhenke** [comment]
@alamb
Thank you so much for the review.

I resolved the merge conflicts and updated the docs according to your comments.

**Comment by alamb** [comment]
Epic work @marvinlanhenke  -- thank you very much!

**Comment by marvinlanhenke** [comment]
... Some epic strg+c+v combos from the CSV implementation @alamb 😂 but thanks again for the kind review. 

I am hoping to improve this implementation next week, by reducing the number of GetRequests on the object store down to one. However, the stream handling is more complicated than I wished for...

**Comment by alamb** [comment]
> ... Some epic strg+c+v combos from the CSV implementation @alamb 😂 but thanks again for the kind review.

Well, I think being able to extract and reuse code is far harder than just copy/paste/modify which you could have done ;) 

> However, the stream handling is more complicated than I wished for...

Yeah -- I agree -- it might need to get wrapped up into its own state machine / stream impl 🤔

**Pull request merged**